### PR TITLE
A01 - Add gitea support

### LINKED ISSUE #1451: Add gitea support
gitea is a highly popular git server written in go, which looks very similar to github.

### PR DESCRIPTION
**Pull request by blaggacao**
closes #1451 

- [x] #1457
- [x] rebase
- [x] fmt / vet
- [x] approved or denied int64 for PR https://github.com/git-town/git-town/issues/1478
- [x] tests
- [x] update docs

### CODE REVIEW

**Comment by kevgo** [comment]
Excellent! Can you please sync this with the latest changes from `master` in your upstream? Also, Git Town uses vendored dependencies. Please run `go mod vendor` and then `go mod tidy` and commit the resulting changes. Thanks!

**Comment by blaggacao** [comment]
Will wait for ongoing changes, indeed my golang setup has become rusty, not used for a while. I quite some pointers are wrongly set as well...

File: src/drivers/gitea.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+type giteaCodeHostingDriver struct {
+	originURL  string
+	hostname   string
+	apiToken   string
+	client     *gitea.Client
+	owner      string
+	repository string
+}
+
+func (d *giteaCodeHostingDriver) HostingServiceName() string {
```
[the remark this answers is not in the document]
Comment (blaggacao) [reply src/drivers/gitea.go]: except for helpers, done in https://github.com/git-town/git-town/pull/1454/commits/cce8d95b9daee48c6abfdb2e9e3247f07e5bd439

**Comment by blaggacao** [comment]
@kevgo Maybe you can point me towards how to get to the offending value or how to get a `fmt.Printf` into that output?

```
[0xc0001db2c0]
--- FAIL: TestGiteaDriver_CanMergePullRequest (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x6cf15b]

goroutine 14 [running]:
testing.tRunner.func1.1(0x73e280, 0xa4d430)
	/usr/local/go/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc000139320)
	/usr/local/go/src/testing/testing.go:943 +0x3f9
panic(0x73e280, 0xa4d430)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/git-town/git-town/src/drivers.filterPullRequests(0xc00007b380, 0x1, 0x32, 0x7a77fc, 0x4, 0xc000129e78, 0xd, 0xc000129e78, 0xd, 0x0)
	/home/blaggacao/go/src/github.com/git-town/git-town/src/drivers/gitea.go:155 +0x15b
github.com/git-town/git-town/src/drivers.(*giteaCodeHostingDriver).CanMergePullRequest(0xc0000604e0, 0x7a8536, 0x7, 0x7a77fc, 0x4, 0xc0001dc3a0, 0x1, 0xc000039748, 0x457979, 0x9f97ee8892a3, ...)
	/home/blaggacao/go/src/github.com/git-town/git-town/src/drivers/gitea.go:46 +0x1a4
github.com/git-town/git-town/src/drivers_test.TestGiteaDriver_CanMergePullRequest(0xc000139320)
	/home/blaggacao/go/src/github.com/git-town/git-town/src/drivers/gitea_test.go:33 +0x136
testing.tRunner(0xc000139320, 0x7c3130)
	/usr/local/go/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1042 +0x357
FAIL	github.com/git-town/git-town/src/drivers	0.011s

```

**Comment by kevgo** [comment]
Hi @blaggacao, you are de-referencing a `nil` value on [line 155](https://github.com/blaggacao/git-town/blob/gitea-support/src/drivers/gitea.go#L154). The culprit is that `pullRequest.Base` in `pullRequest.Base.Name` is `nil`. 

You can see this in the stacktrace that Go provides. Ignore the locations in internal Go libraries. A few lines down it correctly prints `/home/blaggacao/go/src/github.com/git-town/git-town/src/drivers/gitea.go:155`. 

You can add a statement like `fmt.Println(11111111, pullRequest.Base)` there. Then run the tests, look for "11111111" in the output (it should stand out), then you see that the value is indeed `nil`.

A few lines below you see that the failing test is `/home/blaggacao/go/src/github.com/git-town/git-town/src/drivers/gitea_test.go:33`. That's [TestGiteaDriver_CanMergePullRequest](https://github.com/blaggacao/git-town/blob/gitea-support/src/drivers/gitea_test.go#L29-L39). Maybe the mocked HTTP requests need to return more data?

Btw if you happen to use VSCode, you can debug tests super easy right from the IDE. Probably also in other IDEs or even in Vim or Emacs. But I only know how to do it in VSCode. Please let me know if you want some pointers.

**Comment by blaggacao** [comment]
Nice trick: `11111111 <nil>`, thank you. I'm used to post-mortem to be able to dig around in such cases. Is https://github.com/derekparker/delve a viable option to get rolling?

**Comment by kevgo** [comment]
Yes, delve is THE debugger for Go. I don't know how to set breakpoints, run it, and inspect variables from the command line though. This is best done with an IDE like VSCode or Goland. Or just go with debugging statements using `fmt.Println`. You can run just the broken test like so: `go test ./... -run TestGiteaDriver_CanMergePullRequest`

**Comment by blaggacao** [comment]
I can't see details on CI without logging in...

**Comment by blaggacao** [comment]
Maybe failure is due to
```golang
src/drivers/gitea.go:188:113: (*giteaCodeHostingDriver).apiRetargetPullRequests - result 0 (error) is always nil (unparam)
func (d *giteaCodeHostingDriver) apiRetargetPullRequests(pullRequests []*gitea.PullRequest, newBaseName string) error {
```

This is on purpose to be "activated" once upstream improvements are released.

**Comment by blaggacao** [comment]
moved to #1487

**Pull request closed without merging**